File: Design-Patterns-Using-Cpp/ArchitecturalPatterns/PipeAndFilter.cpp
#include <iostream>
#include <string>
#include <vector>

// Filter Base Class
class Filter {
public:
    virtual std::string process(const std::string& data) = 0;
    virtual ~Filter() {}
};

// Filters
class CapitalizeFilter : public Filter {
public:
    std::string process(const std::string& data) override {
        std::string processedData = data;
        for (char& c : processedData) {
            c = std::toupper(c);
        }
        return processedData;
    }
};

class ReplaceSpaceFilter : public Filter {
public:
    std::string process(const std::string& data) override {
        std::string processedData = data;
        for (char& c : processedData) {
            if (c == ' ') {
                c = '_';
            }
        }
        return processedData;
    }
};

class RemoveSpecialCharactersFilter : public Filter {
public:
    std::string process(const std::string& data) override {
        std::string processedData;
        for (char c : data) {
            if (c != ',' && c != '@' && c != '!') {
                processedData += c;
            }
        }
        return processedData;
    }
};

// Data Processing Pipeline
class DataProcessingPipeline {
private:
    std::vector<Filter*> filters;

public:
    ~DataProcessingPipeline() {
        for (Filter* filter : filters) {
            delete filter;
        }
    }

    void addFilter(Filter* filter) {
        filters.push_back(filter);
    }

    std::string processData(const std::string& data) {
        std::string processedData = data;
        for (Filter* filter : filters) {
            processedData = filter->process(processedData);
        }
        return processedData;
    }
};

// Main function
int main() {
    DataProcessingPipeline pipeline;
    pipeline.addFilter(new CapitalizeFilter());
    pipeline.addFilter(new ReplaceSpaceFilter());
    pipeline.addFilter(new RemoveSpecialCharactersFilter());

    std::string data = "Hello, World!";
    std::string result = pipeline.processData(data);
    std::cout << "Result: " << result << std::endl;  // Output: "HELLO_WORLD"

    return 0;
}

/*
Result: HELLO_WORLD
*/

File: Design-Patterns-Using-Swift/ArchitecturalPatterns/PipeAndFilter.swift
import Foundation

// Filter Protocol
protocol Filter {
    func process(data: String) -> String
}

// Filters
class CapitalizeFilter: Filter {
    func process(data: String) -> String {
        return data.uppercased()
    }
}

class ReplaceSpaceFilter: Filter {
    func process(data: String) -> String {
        return data.replacingOccurrences(of: " ", with: "_")
    }
}

class RemoveSpecialCharactersFilter: Filter {
    func process(data: String) -> String {
        let specialCharacters = ",@!"
        var result = ""
        for char in data {
            if !specialCharacters.contains(char) {
                result.append(char)
            }
        }
        return result
    }
}

// Data Processing Pipeline
class DataProcessingPipeline {
    private var filters: [Filter] = []

    func addFilter(_ filter: Filter) {
        filters.append(filter)
    }

    func processData(_ data: String) -> String {
        var processedData = data
        for filter in filters {
            processedData = filter.process(data: processedData)
        }
        return processedData
    }
}

// Client Code
let pipeline = DataProcessingPipeline()
pipeline.addFilter(CapitalizeFilter())
pipeline.addFilter(ReplaceSpaceFilter())
pipeline.addFilter(RemoveSpecialCharactersFilter())

let data = "Hello, World!"
let result = pipeline.processData(data)
print("Result: \(result)") // Output: "HELLO_WORLD"

/*
Result: HELLO_WORLD
*/
